File: nodejs/test/e2e/session_config.test.ts
import { describe, expect, it } from "vitest";
import { writeFile, mkdir } from "fs/promises";
import { join } from "path";
import { approveAll } from "../../src/index.js";
import { createSdkTestContext } from "./harness/sdkTestContext.js";

describe("Session Configuration", async () => {
    const { copilotClient: client, workDir, openAiEndpoint } = await createSdkTestContext();

    it("should use workingDirectory for tool execution", async () => {
        const subDir = join(workDir, "subproject");
        await mkdir(subDir, { recursive: true });
        await writeFile(join(subDir, "marker.txt"), "I am in the subdirectory");

        const session = await client.createSession({
            onPermissionRequest: approveAll,
            workingDirectory: subDir,
        });

        const assistantMessage = await session.sendAndWait({
            prompt: "Read the file marker.txt and tell me what it says",
        });
        expect(assistantMessage?.data.content).toContain("subdirectory");

        await session.disconnect();
    });

    it("should create session with custom provider config", async () => {
        const session = await client.createSession({
            onPermissionRequest: approveAll,
            provider: {
                baseUrl: "https://api.example.com/v1",
                apiKey: "test-key",
            },
        });

        expect(session.sessionId).toMatch(/^[a-f0-9-]+$/);

        try {
            await session.disconnect();
        } catch {
            // disconnect may fail since the provider is fake
        }
    });

    it("should accept blob attachments", async () => {
        // Write the image to disk so the model can view it if it tries
        const pngBase64 =
            "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNk+M9QDwADhgGAWjR9awAAAABJRU5ErkJggg==";
        await writeFile(join(workDir, "pixel.png"), Buffer.from(pngBase64, "base64"));

        const session = await client.createSession({ onPermissionRequest: approveAll });

        await session.sendAndWait({
            prompt: "What color is this pixel? Reply in one word.",
            attachments: [
                {
                    type: "blob",
                    data: pngBase64,
                    mimeType: "image/png",
                    displayName: "pixel.png",
                },
            ],
        });

        await session.disconnect();
    });

    it("should accept message attachments", async () => {
        await writeFile(join(workDir, "attached.txt"), "This file is attached");

        const session = await client.createSession({ onPermissionRequest: approveAll });

        await session.sendAndWait({
            prompt: "Summarize the attached file",
            attachments: [{ type: "file", path: join(workDir, "attached.txt") }],
        });

        await session.disconnect();
    });

    const PNG_1X1 = Buffer.from(
        "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNk+M9QDwADhgGAWjR9awAAAABJRU5ErkJggg==",
        "base64"
    );
    const VIEW_IMAGE_PROMPT =
        "Use the view tool to look at the file test.png and describe what you see";

    function hasImageUrlContent(messages: Array<{ role: string; content: unknown }>): boolean {
        return messages.some(
            (m) =>
                m.role === "user" &&
                Array.isArray(m.content) &&
                m.content.some((p: { type: string }) => p.type === "image_url")
        );
    }

    it("vision disabled then enabled via setModel", async () => {
        await writeFile(join(workDir, "test.png"), PNG_1X1);

        const session = await client.createSession({
            onPermissionRequest: approveAll,
            modelCapabilities: { supports: { vision: false } },
        });

        // Turn 1: vision off — no image_url expected
        await session.sendAndWait({ prompt: VIEW_IMAGE_PROMPT });
        const trafficAfterT1 = await openAiEndpoint.getExchanges();
        const t1Messages = trafficAfterT1.flatMap((e) => e.request.messages ?? []);
        expect(hasImageUrlContent(t1Messages)).toBe(false);

        // Switch vision on (re-specify same model with updated capabilities)
        await session.setModel("claude-sonnet-4.5", {
            modelCapabilities: { supports: { vision: true } },
        });

        // Turn 2: vision on — image_url expected
        await session.sendAndWait({ prompt: VIEW_IMAGE_PROMPT });
        const trafficAfterT2 = await openAiEndpoint.getExchanges();
        // Only check exchanges added after turn 1
        const newExchanges = trafficAfterT2.slice(trafficAfterT1.length);
        const t2Messages = newExchanges.flatMap((e) => e.request.messages ?? []);
        expect(hasImageUrlContent(t2Messages)).toBe(true);

        await session.disconnect();
    });

    it("vision enabled then disabled via setModel", async () => {
        await writeFile(join(workDir, "test.png"), PNG_1X1);

        const session = await client.createSession({
            onPermissionRequest: approveAll,
            modelCapabilities: { supports: { vision: true } },
        });

        // Turn 1: vision on — image_url expected
        await session.sendAndWait({ prompt: VIEW_IMAGE_PROMPT });
        const trafficAfterT1 = await openAiEndpoint.getExchanges();
        const t1Messages = trafficAfterT1.flatMap((e) => e.request.messages ?? []);
        expect(hasImageUrlContent(t1Messages)).toBe(true);

        // Switch vision off
        await session.setModel("claude-sonnet-4.5", {
            modelCapabilities: { supports: { vision: false } },
        });

        // Turn 2: vision off — no image_url expected in new exchanges
        await session.sendAndWait({ prompt: VIEW_IMAGE_PROMPT });
        const trafficAfterT2 = await openAiEndpoint.getExchanges();
        const newExchanges = trafficAfterT2.slice(trafficAfterT1.length);
        const t2Messages = newExchanges.flatMap((e) => e.request.messages ?? []);
        expect(hasImageUrlContent(t2Messages)).toBe(false);

        await session.disconnect();
    });
});

File: go/internal/e2e/session_config_test.go
package e2e

import (
	"encoding/base64"
	"encoding/json"
	"os"
	"path/filepath"
	"testing"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
)

// hasImageURLContent returns true if any user message in the given exchanges
// contains an image_url content part (multimodal vision content).
func hasImageURLContent(exchanges []testharness.ParsedHttpExchange) bool {
	for _, ex := range exchanges {
		for _, msg := range ex.Request.Messages {
			if msg.Role == "user" && len(msg.RawContent) > 0 {
				var content []interface{}
				if json.Unmarshal(msg.RawContent, &content) == nil {
					for _, part := range content {
						if m, ok := part.(map[string]interface{}); ok {
							if m["type"] == "image_url" {
								return true
							}
						}
					}
				}
			}
		}
	}
	return false
}

func TestSessionConfig(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	client := ctx.NewClient()
	t.Cleanup(func() { client.ForceStop() })

	if err := client.Start(t.Context()); err != nil {
		t.Fatalf("Failed to start client: %v", err)
	}

	// Write 1x1 PNG to the work directory
	png1x1, err := base64.StdEncoding.DecodeString("iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNk+M9QDwADhgGAWjR9awAAAABJRU5ErkJggg==")
	if err != nil {
		t.Fatalf("Failed to decode PNG: %v", err)
	}
	if err := os.WriteFile(filepath.Join(ctx.WorkDir, "test.png"), png1x1, 0644); err != nil {
		t.Fatalf("Failed to write test.png: %v", err)
	}

	viewImagePrompt := "Use the view tool to look at the file test.png and describe what you see"

	t.Run("vision disabled then enabled via setModel", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			ModelCapabilities: &copilot.ModelCapabilitiesOverride{
				Supports: &copilot.ModelCapabilitiesOverrideSupports{
					Vision: copilot.Bool(false),
				},
			},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		// Turn 1: vision off — no image_url expected
		if _, err := session.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: viewImagePrompt}); err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		trafficAfterT1, err := ctx.GetExchanges()
		if err != nil {
			t.Fatalf("Failed to get exchanges: %v", err)
		}
		if hasImageURLContent(trafficAfterT1) {
			t.Error("Expected no image_url content parts when vision is disabled")
		}

		// Switch vision on
		if err := session.SetModel(t.Context(), "claude-sonnet-4.5", &copilot.SetModelOptions{
			ModelCapabilities: &copilot.ModelCapabilitiesOverride{
				Supports: &copilot.ModelCapabilitiesOverrideSupports{
					Vision: copilot.Bool(true),
				},
			},
		}); err != nil {
			t.Fatalf("SetModel returned error: %v", err)
		}

		// Turn 2: vision on — image_url expected in new exchanges
		if _, err := session.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: viewImagePrompt}); err != nil {
			t.Fatalf("Failed to send second message: %v", err)
		}

		trafficAfterT2, err := ctx.GetExchanges()
		if err != nil {
			t.Fatalf("Failed to get exchanges after turn 2: %v", err)
		}
		newExchanges := trafficAfterT2[len(trafficAfterT1):]
		if !hasImageURLContent(newExchanges) {
			t.Error("Expected image_url content parts when vision is enabled")
		}
	})

	t.Run("vision enabled then disabled via setModel", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			ModelCapabilities: &copilot.ModelCapabilitiesOverride{
				Supports: &copilot.ModelCapabilitiesOverrideSupports{
					Vision: copilot.Bool(true),
				},
			},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		// Turn 1: vision on — image_url expected
		if _, err := session.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: viewImagePrompt}); err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		trafficAfterT1, err := ctx.GetExchanges()
		if err != nil {
			t.Fatalf("Failed to get exchanges: %v", err)
		}
		if !hasImageURLContent(trafficAfterT1) {
			t.Error("Expected image_url content parts when vision is enabled")
		}

		// Switch vision off
		if err := session.SetModel(t.Context(), "claude-sonnet-4.5", &copilot.SetModelOptions{
			ModelCapabilities: &copilot.ModelCapabilitiesOverride{
				Supports: &copilot.ModelCapabilitiesOverrideSupports{
					Vision: copilot.Bool(false),
				},
			},
		}); err != nil {
			t.Fatalf("SetModel returned error: %v", err)
		}

		// Turn 2: vision off — no image_url expected in new exchanges
		if _, err := session.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: viewImagePrompt}); err != nil {
			t.Fatalf("Failed to send second message: %v", err)
		}

		trafficAfterT2, err := ctx.GetExchanges()
		if err != nil {
			t.Fatalf("Failed to get exchanges after turn 2: %v", err)
		}
		newExchanges := trafficAfterT2[len(trafficAfterT1):]
		if hasImageURLContent(newExchanges) {
			t.Error("Expected no image_url content parts when vision is disabled")
		}
	})
}
